File: src/main/scala/com/wikibooks/spark/ch6/scala/KafkaSample.scala
package com.wikibooks.spark.ch6.scala

import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.common.serialization.StringDeserializer
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.streaming.Seconds
import org.apache.spark.streaming.StreamingContext
import org.apache.spark.streaming.kafka010.ConsumerStrategies.Subscribe
import org.apache.spark.streaming.kafka010.KafkaUtils
import org.apache.spark.streaming.kafka010.LocationStrategies.PreferConsistent

// 6.2.4절 예제 6-10
object KafkaSample extends Serializable {

  def main(args: Array[String]) {

    val conf = new SparkConf()
      .setMaster("local[*]")
      .setAppName("KafkaSample")

    val sc = new SparkContext(conf)
    val ssc = new StreamingContext(sc, Seconds(3))

    val params = Map(
      ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG -> "localhost:9092",
      ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG -> classOf[StringDeserializer],
      ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG -> classOf[StringDeserializer],
      ConsumerConfig.GROUP_ID_CONFIG -> "test-group-1",
      ConsumerConfig.AUTO_OFFSET_RESET_CONFIG -> "latest")

    val topics = Array("test")

    val ds = KafkaUtils.createDirectStream[String, String](
      ssc,
      PreferConsistent,
      Subscribe[String, String](topics, params))

    ds.flatMap(record => record.value.split(" "))
      .map((_, 1))
      .reduceByKey(_ + _)
      .print

    ssc.start
    ssc.awaitTermination()
  }
}

File: src/main/java/com/wikibooks/spark/ch6/KafkaSample.java
package com.wikibooks.spark.ch6;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.apache.spark.streaming.kafka010.ConsumerStrategies;
import org.apache.spark.streaming.kafka010.KafkaUtils;
import org.apache.spark.streaming.kafka010.LocationStrategies;

import scala.Tuple2;

public class KafkaSample {

  public static void main(String[] args) throws Exception {

    SparkConf conf = new SparkConf().setMaster("local[*]").setAppName("KafkaSample");
    JavaSparkContext sc = new JavaSparkContext(conf);
    JavaStreamingContext ssc = new JavaStreamingContext(sc, Durations.seconds(3));

    Map<String, Object> params = new HashMap<>();
    params.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
    params.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
    params.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
    params.put(ConsumerConfig.GROUP_ID_CONFIG, "test-group-2");
    params.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

    List<String> topics = Arrays.asList("test");

    JavaInputDStream<ConsumerRecord<String, String>> ds = KafkaUtils.createDirectStream(ssc,
            LocationStrategies.PreferConsistent(),
            ConsumerStrategies.<String, String>Subscribe(topics, params));
    
    ds.flatMap((ConsumerRecord<String, String>  record) -> Arrays.asList(record.value().split(" ")).iterator())
    .mapToPair((String word) -> new Tuple2<String, Integer>(word, 1))
    .reduceByKey((Integer v1, Integer v2) -> v1 + v2)
    .print();
    
    ssc.start();
    ssc.awaitTermination();
  }
}